Fix CPU/CUDA inconsistency in cosine_similarity for large eps

vishalgoyal316 opened the pull request:
- Add validation: eps must be non-negative
- Convert eps to scalar tensor for consistent overflow handling
- Use higher precision dtype between norms to avoid unnecessary overflow
- Update tests to validate negative eps and overflow cases

When eps overflows the input dtype (e.g., eps=1e23 with float16 inputs), CPU previously threw an error while CUDA silently produced NaN. Now CPU handle overflow consistent with CUDA by converting eps to a tensor with appropriate dtype, allowing graceful overflow to inf.

Fixes #153209

vishalgoyal316 committed:
Fix CPU/CUDA inconsistency in cosine_similarity for large eps

- Add validation: eps must be non-negative
- Convert eps to scalar tensor for consistent overflow handling
- Use higher precision dtype between norms to avoid unnecessary overflow
- Update tests to validate negative eps and overflow cases

When eps overflows the input dtype (e.g., eps=1e23 with float16 inputs),
CPU previously threw an error while CUDA silently produced NaN. Now both
devices handle overflow consistently by converting eps to a tensor with
appropriate dtype, allowing graceful overflow to inf.

Fixes #153209

vishalgoyal316 commented:
CC: @abhitorch81 @cleonard530 @ahkush @gaurav-redhat

vishalgoyal316 commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -320,10 +322,18 @@ Tensor cosine_similarity(const Tensor& x1_, const Tensor& x2_, int64_t dim, doub
+  auto eps_options = at::isReducedFloatingType(x1_norm.scalar_type()) &&
+                     !at::isReducedFloatingType(x2_norm.scalar_type())
+                     ? x2_norm.options() : x1_norm.options();
+  auto eps_tensor = at::scalar_tensor(eps, eps_options);
```
Right, and thanks for suggestion. I will try this.

vishalgoyal316 committed:
Optimize cosine_similarity eps dtype selection for CPU/CUDA consistency

Fix incomplete mixed-precision logic that failed when both norms are
reduced floating types (float16/bfloat16). Previous code only checked
one direction, causing CUDA overflow errors with float16 inputs.

Changes:
- Handle all cases: both reduced → use float32, mixed → use non-reduced type
- Cache scalar_type() and isReducedFloatingType() calls (12→4 function calls)
- Ensure consistent CPU/CUDA behavior (both convert overflow to inf vs CUDA error)

Fixes: F.cosine_similarity(float16, float16, eps=large_value) on CUDA

isuruf commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -320,10 +322,25 @@ Tensor cosine_similarity(const Tensor& x1_, const Tensor& x2_, int64_t dim, doub
+  auto eps_dtype = (x1_is_reduced && x2_is_reduced) ? at::kFloat
+                   : x1_is_reduced ? x2_dtype
+                   : x2_is_reduced ? x1_dtype
+                   : at::promoteTypes(x1_dtype, x2_dtype);
```
```suggestion
  auto common_is_reduced = at::isReducedFloatingType(commonDtype);
  auto eps_dtype = common_is_reduced ? at::kFloat : commonDtype;
```

vishalgoyal316 committed:
Update aten/src/ATen/native/Distance.cpp

Co-authored-by: Isuru Fernando <isuruf@gmail.com>

vishalgoyal316 committed:
Update comment in Distance.cpp to clarify eps dtype selection

Simplify comment to accurately reflect the actual code logic that uses float32 for reduced types, otherwise uses commonDtype.

vishalgoyal316 commented:
@pytorchbot merge

vishalgoyal316 commented:
@pytorchbot merge
